### DISCUSSION TITLE
Generating a good graph embedding - seeking "tricks-of-the-trade"

### DISCUSSION THREAD

**Discussion by batlike** [pyg-team/pytorch_geometric#5468] [Q&A] [answered]
As my first foray into graph neural networks, I followed along the [node classification tutorial](https://colab.research.google.com/drive/14OvFnAXggxB8vM4e8vSURUp1TaKnovzX) which I found quite fun! I thought I'd apply what I learned by clustering **graphs** in place of nodes. 

I downloaded the `ENZYMES` dataset from `torch_geometric.datasets.TUDataset` which has 600 graphs - each representing an enzyme molecule. Nodes have 3 features and graphs fall into 6 classes. An initial 2-D graph embedding cannot separate these 6 classes:
![image](https://user-images.githubusercontent.com/5013556/190885819-f0e9f71c-aac7-4587-b530-c2a136558bd5.png)

Using a very simple GCN, I found that my model quickly began to overfit, despite playing around with dropout and learning rates:
![image](https://user-images.githubusercontent.com/5013556/190885835-c4e40066-992c-4cf3-a819-b244986d7d49.png)

The final embedding looks promising:
![image](https://user-images.githubusercontent.com/5013556/190885965-35fa8b03-9ed0-4384-adc5-a453a6a3529f.png)

... but the validation accuracy plateaus at ~40%. Would love to hear what others have tried to (A) avoid overfitting and (B) "tricks" to generate better graph embeddings (are there particular GNNs that are better suited for this task?). I'd love to build some intuition for ie. which GNNs are better for ie. graphs with rich node features, which GNNs are better for rich edges, etc. 

---
```
import torch.nn.functional as F
from torch_geometric.nn import GCNConv, GraphConv
from torch_geometric.nn.norm import BatchNorm
from torch_geometric.nn import global_mean_pool

class GCN(torch.nn.Module):
    
    def __init__(self, n_hidden):
        super(GCN, self).__init__()
        torch.manual_seed(12345)
        self.norm1 = BatchNorm(dataset.num_node_features)
        self.norm2 = BatchNorm(n_hidden)
        self.norm3 = BatchNorm(n_hidden)
        self.conv1 = GraphConv(dataset.num_node_features, n_hidden)
        self.conv2 = GraphConv(n_hidden, n_hidden)
        self.conv3 = GraphConv(n_hidden, n_hidden)
        self.lin = Linear(n_hidden, dataset.num_classes)
        
    def forward(self, x, edge_index, batch):
        
        x = self.norm1(x)
        # gather node embeddings
        x = self.conv1(x, edge_index) 
        x = x.relu()
        x = self.norm2(x)
        x = self.conv2(x, edge_index)
        x = x.relu()
        x = self.norm3(x)
        x = self.conv3(x, edge_index)
        
        # compute unifed graph embedding (readout)
        x = global_mean_pool(x, batch)
        
        # classify (nonlinear activation)
        x = F.dropout(x, p=0.5, training=self.training)
        x = self.lin(x)
        return x
```

**Comment by EdisonLeeeee** [comment, marked as answer]
Hi @batlike could you show me the code you are using for training? I used the code you provided to define the model and I got ~60% validation accuracy. BTW, this model is not GCN. In PyG, GCN is defined by `GCNConv` layers.

**Reply by batlike** [reply]
Thanks for this feedback @EdisonLeeeee ! You are right, I had initially used `GCNConv` Layers and tried playing with other types of layers to see if it improves results. I am still trying to build intuition for which layers are best suited for different types of graphs. 

Here is how I trained my network:
```
# split
train_dataset = dataset[:int(0.8*len(dataset))] 
test_dataset = dataset[int(0.8*len(dataset)):]

# data loader
train_loader = DataLoader(train_dataset, batch_size = 50, shuffle = True)
test_loader = DataLoader(test_dataset, batch_size = 50, shuffle = False)

model = GCN(n_hidden = 20)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
criterion = torch.nn.CrossEntropyLoss()

def train():
    model.train()
    
    for data in train_loader:
        out = model(data.x, data.edge_index, data.batch)
        loss = criterion(out, data.y) # compute loss
        loss.backward() # calculate gradients
        optimizer.step() # update parameters given gradient
        optimizer.zero_grad() # clear gradients
        
def test(loader):
    model.eval()
    
    correct = 0
    for data in loader:
        out = model(data.x, data.edge_index, data.batch)
        pred = out.argmax(dim=1)
        correct += int((pred == data.y).sum())
    return correct / len(loader.dataset)

train_res, test_res = [], []
for epoch in range(1, 1000):
    train()
    train_acc = test(train_loader)
    test_acc = test(test_loader)
    
    train_res.append(train_acc)
    test_res.append(test_acc)
    
    print(f'Epoch: {epoch:03d}, Train Acc: {train_acc:.4f}, Test Acc: {test_acc:.4f}')
```

**Reply by EdisonLeeeee** [reply]
Your code looks fine. In `ENZYMES` dataset, you might need to set `use_node_attr=True` to add node attributes into the dataset:
```python
dataset = TUDataset("data/", name="ENZYMES", use_node_attr=True)
```
This can increase the accuracy up to ~60%.

> which layers are best suited for different types of graphs.

As far as I know, there is no layer that performs best for different types of graphs. If it must have one, I would vote for `GINConv` or `TransformerConv`. IMO, a better pooling layer is can bring more benefits for graph-level tasks. You can take a look at this [benchmark](https://paperswithcode.com/paper/bridging-the-gap-between-spectral-and-spatial#code).

**Reply by batlike** [reply]
Thank you for linking to that benchmark! 

In case others are also looking for building intuition on which layers are best for types of graphs, I found this post helpful:
https://github.com/pyg-team/pytorch_geometric/discussions/2549

Maybe others can build on this list, but according to @rusty1s they are:
- graph structure important --> `GIN`, `PNAConv`
- features important --> `GATConv`, `TransformerConv`
- requirement to smooth features --> `GCNConv`, `SAGEConv`
- edge features important --> `NNConv`, `GMMConv`
- hierarchical graphs --> `DiffPool`

Actually, I wonder for hierarchical graphs if there are GNNs adapted for hyperbolic embeddings.
